### PR TITLE
[operator] Manage `gardener-metrics-exporter`

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:
With this PR, `gardener-operator` is enhanced to manage `gardener-metrics-exporter`.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/7016

**Special notes for your reviewer**:
~~In draft until https://github.com/gardener/gardener/pull/8309 this is merged.~~

**Release note**:

```feature operator
`gardener-operator` now takes over management of `gardener-metrics-exporter`.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
FYI @istvanballok @rickardsjp

File: pkg/operator/controller/garden/care/health_test.go
Comment (dimitar-kostadinov) [inline pkg/operator/controller/garden/care/health_test.go:690]: Rather than replacing the previous assertion the new one should be added.

**Review by dimitar-kostadinov — COMMENTED** [review]
While checking it locally:
The `garden` OBSERVABILITY condition is false, because `gardener-metrics-exporter` is not ready:
```
  - lastTransitionTime: "2023-09-05T08:25:00Z"
    lastUpdateTime: "2023-09-05T08:20:59Z"
    message: 'Missing required deployments: [gardener-metrics-exporter]'
    reason: DeploymentMissing
    status: "False"
    type: ObservabilityComponentsHealthy
```

The root cause is:
```
W0905 08:52:47.475401       1 reflector.go:424] pkg/mod/k8s.io/client-go@v0.25.0/tools/cache/reflector.go:169: failed to list *v1beta1.Project: the server could not find the requested resource (get projects.core.gardener.cloud)
E0905 08:52:47.475667       1 reflector.go:140] pkg/mod/k8s.io/client-go@v0.25.0/tools/cache/reflector.go:169: Failed to watch *v1beta1.Project: failed to list *v1beta1.Project: the server could not find the requested resource (get projects.core.gardener.cloud)
W0905 08:52:52.006258       1 reflector.go:424] pkg/mod/k8s.io/client-go@v0.25.0/tools/cache/reflector.go:169: failed to list *v1beta1.Seed: the server could not find the requested resource (get seeds.core.gardener.cloud)
E0905 08:52:52.006646       1 reflector.go:140] pkg/mod/k8s.io/client-go@v0.25.0/tools/cache/reflector.go:169: Failed to watch *v1beta1.Seed: failed to list *v1beta1.Seed: the server could not find the requested resource (get seeds.core.gardener.cloud)
time="2023-09-05T08:52:52Z" level=info msg="Received interupt signal."
time="2023-09-05T08:52:52Z" level=error msg="Timed out waiting for Garden caches to sync"
```

File: imagevector/images.yaml
Comment (rfranzke) [inline imagevector/images.yaml]: Please move this somewhere below L162 - we only enlist the core Gardener images here

File: imagevector/images.yaml
Comment (rfranzke) [inline imagevector/images.yaml]: I don't think this is necessary here

File: pkg/component/gardenermetricsexporter/gardener_metrics_exporter_suite_test.go
Comment (rfranzke) [inline pkg/component/gardenermetricsexporter/gardener_metrics_exporter_suite_test.go]:
```suggestion
func TestGardenerMetricsExporter(t *testing.T) {
```

File: pkg/operator/controller/garden/garden/reconciler_reconcile.go
Comment (rfranzke) [inline pkg/operator/controller/garden/garden/reconciler_reconcile.go]: 1. We should also wait for the `gardener-apiserver` here after #8309 is merged, right?
1. Why is the KSM deployment task above not depending on `waitUntilKubeAPIServerIsReady` as well?

File: test/integration/operator/garden/care/care_test.go
Comment (rfranzke) [inline test/integration/operator/garden/care/care_test.go:95]: You could also adapt https://github.com/gardener/gardener/blob/master/test/integration/operator/garden/garden/garden_test.go

File: pkg/component/gardenermetricsexporter/service.go
Comment (rfranzke) [inline pkg/component/gardenermetricsexporter/service.go]:
```suggestion
	return &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      serviceName,
			Namespace: g.namespace,
			Labels:    GetLabels(),
		},
		Spec: corev1.ServiceSpec{
			Type:            corev1.ServiceTypeClusterIP,
			SessionAffinity: corev1.ServiceAffinityNone,
			Selector:        GetLabels(),
			Ports: []corev1.ServicePort{
				{
					Protocol:   corev1.ProtocolTCP,
					Port:       probePort,
					TargetPort: intstr.FromInt(probePort),
				},
			},
		},
	}
```

File: pkg/component/gardenermetricsexporter/gardener_metrics_exporter.go
Comment (rfranzke) [inline pkg/component/gardenermetricsexporter/gardener_metrics_exporter.go]:
```suggestion
	virtualRegistry := managedresources.NewRegistry(operatorclient.VirtualScheme, operatorclient.VirtualCodec, operatorclient.VirtualSerializer)
```

File: pkg/component/gardenermetricsexporter/deployment.go
Comment (rfranzke) [inline pkg/component/gardenermetricsexporter/deployment.go]: Set the `revisionHistoryLimit` to `2`

**Review by rfranzke — CHANGES_REQUESTED** [review]
Nice :)

**Review by rfranzke — COMMENTED** [review]
Also adapt https://github.com/gardener/gardener/blob/master/docs/concepts/operator.md#main-reconciler pls

**Comment by rfranzke** [comment]
#8309 is ready, please address the comments and rebase :)

File: pkg/operator/controller/garden/garden/reconciler_reconcile.go
Comment (acumino) [reply pkg/operator/controller/garden/garden/reconciler_reconcile.go]: > Why is the KSM deployment task above not depending on waitUntilKubeAPIServerIsReady as well?

Not sure. I will add `waitUntilKubeAPIServerIsReady` as a dependency for KSM.

File: pkg/operator/controller/garden/care/health_test.go
Comment (acumino) [reply pkg/operator/controller/garden/care/health_test.go:690]: Error returns only one deployment at a time. That's why there is only one error.

File: imagevector/images.yaml
Comment (rfranzke) [inline imagevector/images.yaml]: remove this

```suggestion
```

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

File: imagevector/images.yaml
Comment (acumino) [reply imagevector/images.yaml]: In rebase it moved up 🤦 , while removing conflict

**Pull request merged**
